descartes_planner/src/planning_graph.cpp

```diff
@@ -36,16 +36,108 @@
 
+#include <future>
+
 using namespace descartes_core;
 using namespace descartes_trajectory;
+
+namespace
+{
+  using JointSolutions = std::vector<std::vector<double>>;
+  using VecJointSolutions = std::vector<JointSolutions>;
+  using ConstTrajectoryIterator = std::vector<descartes_core::TrajectoryPtPtr>::const_iterator;
+
+  using JointSolutionResult = std::pair<VecJointSolutions, bool>; // result and flag if everything went well
+
+  JointSolutionResult
+  rangeCalculateJointSolutions(descartes_core::RobotModelPtr model, 
+                               ConstTrajectoryIterator begin,
+                               ConstTrajectoryIterator end)
+  {
+    JointSolutionResult solutions;
+    solutions.first.reserve(std::distance(begin, end));
+    solutions.second = true;
+
+    while (begin != end)
+    {
+      JointSolutions joint_poses;
+      begin->get()->getJointPoses(*model, joint_poses);
+
+      if (joint_poses.empty())
+      {
+        solutions.second = false;
+        ROS_ERROR("No joint solutions available for point %s", 
+          boost::uuids::to_string(begin->get()->getID()).c_str());
+      }
+
+      solutions.first.push_back(std::move(joint_poses));
+      ++begin;
+    }
+
+    return solutions;
+  }
+
+  JointSolutionResult
+  parallelCalculateJointSolutions(const std::vector<descartes_core::TrajectoryPtPtr>& trajectory,
+                                  const descartes_core::RobotModel& model,
+                                  const unsigned max_threads)
+  {
+    // Might want to insert check to see if work to be done is small and if so,
+    // just run the IKs synchronously
+    JointSolutionResult solutions;
+    solutions.first.reserve(trajectory.size());
+    solutions.second = true; // success by default
+
+    // Divide work into chunks
+    const unsigned chunk_size = trajectory.size() / max_threads;
+    const unsigned chunk_extra = trajectory.size() % max_threads;
+
+    // Create tasks
+    std::vector<std::future<JointSolutionResult>> futures (max_threads);
+    for (size_t i = 0; i < futures.size(); ++i)
+    {
+      auto start = trajectory.cbegin() + i * chunk_size;
+      auto end = start + chunk_size;
+
+      // Give extra work to last element
+      if (i == futures.size() - 1) end += chunk_extra;
+
+      // Create a seperate copy of the robot state for this worker
+      descartes_core::RobotModelPtr state_copy = model.clone();
+      // Launch task
+      futures[i] = std::async(std::launch::async, rangeCalculateJointSolutions, state_copy, start, end);
+    }
+
+    // Wait for solutions
+    for (auto& future : futures)
+    {
+      JointSolutionResult solution_segment = std::move(future.get());
+
+      solutions.second = solutions.second && solution_segment.second; // carry the false flag forward if is returned
+      if (!solution_segment.second)
+      {
+        ROS_ERROR("One segment of the calculation failed to find a solution for all points");
```

Comment:
In most cases, it'd be preferable to have the planner return as soon as a planning error is encountered. 
For those more select cases where detailed failure information is needed, we could configure the planner to operate in a "debugging/troubleshooting" mode.  This could be accomplished by passing one or more configuration parameters/flags that would be used to set such a mode.

Handling the second more involved case is perhaps outside of the scope of this PR, so it should probably be addressed separately. For now, I suggest that we make the planner quit immediately when a failure is encounter.